### PR TITLE
Type annotate and refactor vulcan-forms

### PR DESCRIPTION
**Pull request by jimrandomh**
Adding type annotations to `vulcan-forms` is trickier than annotating most components, because it uses a lot of props-spreading between wrappers, collection schemas, and dynamically selected form components. This PR gets most of the way to fully type annotating it, setting up interfaces that extend each other to represent prop-spreading, where appropriate.

It also refactors `FormWrapper` to use hooks, rather than dynamically assembling HoCs, which was both an obstacle to type-annotation and also was the last usage left in the codebase for some of those HoCs.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1204312765612282) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/components/vulcan-forms/Form.tsx [hunk 1]
```diff
@@ -128,22 +101,28 @@ const getInitialStateFromProps = nextProps => {
...
 
 /**
- * Note: Only use this through WrappedSmartForm
+ * Main form component. Should not be used directly; use only through
+ * WrappedSmartForm. Depending whether this is a new form or an edit form,
+ * FormWrapper will have provided a create mutator, document and/or update
+ * mutator.
+ *
+ * This is not in the Components table and is not registered with
+ * registerComponent because you aren't supposed to use it without FormWrapper.
```
Comment (jpaddison3) [inline packages/lesswrong/components/vulcan-forms/Form.tsx:122]: ❤️

File: packages/lesswrong/components/vulcan-forms/Form.tsx [hunk 2]
```diff
@@ -382,7 +336,7 @@ class Form<T extends DbObject> extends Component<any,any> {
       document: this.state.initialDocument,
       name: fieldName,
       datatype: fieldSchema.type,
-      layout: this.props.layout,
+      layout: this.props.layout || "horizontal",
```
Comment (jpaddison3) [inline packages/lesswrong/components/vulcan-forms/Form.tsx:339]: That sure is a decision.

(I assume you're maintaining back-compat)

**Review by jpaddison3 — COMMENTED** [review]
Amazing, this is so good. Thanks so much Jim.

(This is not a review, just a glance through the code, looking at the interesting bits.)

**Comment by jpaddison3** [comment]
NB: Test failures are real.

File: packages/lesswrong/components/vulcan-forms/Form.tsx (hunk 2 above)
Comment (jimrandomh) [reply packages/lesswrong/components/vulcan-forms/Form.tsx:339]: This default was already present, using `(Form as any).defaultProps` at the bottom of the file. I'm not sure what this `layout` is used for or if it's even still used.

Comment (jpaddison3) [reply packages/lesswrong/components/vulcan-forms/Form.tsx:339]: I assume it's like, does the Form lay it's components out vertically or horizontally, and someone chose the wrong default according to me.

**Comment by jimrandomh** [comment]
Unit test failure was real; in splitting a component in two, I wound up with a component using `classes` which didn't get styles in `registerComponent`. Oops. Will give this a bit more testing before merging.

File: packages/lesswrong/components/form-components/SocialPreviewUpload.tsx [hunk 3]
```diff
@@ -144,7 +144,7 @@ const buildPreviewFromDocument = (document: PostsEditWithLocalData): {descriptio
 const SocialPreviewUpload = ({name, document, updateCurrentValues, clearField, label, croppingAspectRatio, classes}: {
   name: string,
   document: PostsEditWithLocalData,
-  updateCurrentValues: Function,
+  updateCurrentValues: UpdateCurrentValues,
```
Comment (Will-Howard) [inline packages/lesswrong/components/form-components/SocialPreviewUpload.tsx:147]: Nice, I like this

File: packages/lesswrong/components/vulcan-forms/propTypes.ts [hunk 4]
```diff
@@ -44,3 +44,129 @@ export const callbackProps = {
...
+    currentValues?: any
+    deletedValues?: any
+    errors?: any[]
+    formType: "edit"|"new"
+  }
+  interface FormComponentContext<T> {
+    updateCurrentValues: UpdateCurrentValues
+    addToDeletedValues: any
+  }
+}
```
Comment (Will-Howard) [inline packages/lesswrong/components/vulcan-forms/propTypes.ts:172]: beautiful

**Review by Will-Howard — APPROVED** [review]
Thanks for doing this Jim!

**Pull request merged**